arch/rv32i/src/pmp.rs

```diff
@@ -478,148 +492,164 @@ impl kernel::mpu::MPU for PMPConfig {
-
-                            // Set access to end address
-                            csr::CSR.pmpcfg0.set(cfg_val << 24 | csr::CSR.pmpcfg0.get());
-                            csr::CSR.pmpaddr3.set((start as u32 + size as u32) >> 2);
-                        }
-                        2 => {
-                            // Disable access up to the start address
-                            csr::CSR.pmpcfg1.modify(
-                                csr::pmpconfig::pmpcfg::r0::CLEAR
-                                    + csr::pmpconfig::pmpcfg::w0::CLEAR
-                                    + csr::pmpconfig::pmpcfg::x0::CLEAR
-                                    + csr::pmpconfig::pmpcfg::a0::TOR,
-                            );
-                            csr::CSR.pmpaddr4.set((start as u32) >> 2);
-
-                            // Set access to end address
-                            csr::CSR.pmpcfg1.set(cfg_val << 8 | csr::CSR.pmpcfg0.get());
-                            csr::CSR.pmpaddr5.set((start as u32 + size as u32) >> 2);
-                        }
-                        3 => {
-                            // Disable access up to the start address
-                            csr::CSR.pmpcfg1.modify(
-                                csr::pmpconfig::pmpcfg::r3::CLEAR
-                                    + csr::pmpconfig::pmpcfg::w3::CLEAR
-                                    + csr::pmpconfig::pmpcfg::x3::CLEAR
-                                    + csr::pmpconfig::pmpcfg::a3::TOR,
-                            );
-                            csr::CSR.pmpaddr6.set((start as u32) >> 2);
-
-                            // Set access to end address
-                            csr::CSR.pmpcfg1.set(cfg_val << 24 | csr::CSR.pmpcfg0.get());
-                            csr::CSR.pmpaddr7.set((start as u32 + size as u32) >> 2);
+    fn configure_mpu(&self, config: &Self::MpuConfig, app_id: &AppId) {
+        let last_configured_for_this_app = self
+            .last_configured_for
+            .map_or(false, |last_app_id| last_app_id == app_id);
+
+        // Skip PMP configuration if it is already configured for this app and the MPU
+        // configuration of this app has not changed.
+        if !last_configured_for_this_app || config.is_dirty.get() {
+            // tock/tock#1860
+            let mut regions_sorted = config.regions.clone();
+            regions_sorted.sort_unstable_by(|a, b| {
+                let (a_start, _a_size) = match a.location() {
+                    Some((start, size)) => (start as usize, size),
+                    None => (0xFFFF_FFFF, 0xFFFF_FFFF),
+                };
+                let (b_start, _b_size) = match b.location() {
+                    Some((start, size)) => (start as usize, size),
+                    None => (0xFFFF_FFFF, 0xFFFF_FFFF),
+                };
+                a_start.cmp(&b_start)
+            });
+
+            // tock/tock#1860
+            self.clear_mpu_registers();
```

Comment:
As I mentioned in #1860, I don't think we need to clear the PMP registers when returning to the kernel since we don't use the lock bit. Calling this function here just to stay as close as possible to the current implementation, was planning to remove it as part of #1860.

Reply:
This is confusing then. Can we leave the current `disable_mpu()` as is and work on fixing that? There is already a PR for this: https://github.com/tock/tock/pull/1873